Revert to single arch-specific package (Fix #22)

joshmoore opened the pull request:
~Previously, a special patch was needed on Windows to handle the long jar list. With this patch removed, the main build can again be architecture-neutral.~

Updated `conda build` no longer handles the split packages (reason unknown). This temporarily merges them back into a single build. If anyone has an idea of how to separate them back out, happen to see it!

The original motivation for changing the conda setup task was the lack of support in s-wiegand for the new Mac M! runner.

cc: @jni @jburel @dgault

joshmoore committed:
Set main build to noarch (Fix #22)

Previously, a special patch was needed on Windows
to handle the long jar list. With this patch removed,
the main build can again be architecture-neutral.

manics commented:
If you wanted to test this fully (bearing in mind https://github.com/ome/conda-bioformats2raw/issues/22#issuecomment-1845007375) you'd need to build this under a single architecture, upload it as an artifact, then download and test the built package on multiple architectures.

jni commented:
Kind ping... 🙏

joshmoore commented:
The ping is appreciated, but unfortunately, I think this still isn't going to work.

*Either* this PR needs a way to launch on Windows & Posix without separate .bat & .sh files, *or* we need an OSX/arm runner (and nothing really has changed on that front).

Unless I'm missing something, anyone?

manics commented:
That sounds right. The `bld.bat` and `build.sh` in https://github.com/ome/conda-bioformats2raw/tree/master/bioformats2raw are automatically called depending on the build platform (`bld.bat` when built on Windows, `build.sh` when built on Linux or OS X)- this is why the built packages are different. If it's only built on one platform (as implied by noarch) the resulting package will need to contain both Windows and Linux startup scripts.

If you wanted to be really thorough with testing (not saying you need to be!) you could build it on Linux, [upload the artifact](https://github.com/actions/upload-artifact), then have a follow-on matrix job that runs on all three platforms, [downloads the artifact](https://github.com/actions/download-artifact), and tests it.

jni commented:
> or we need an OSX/arm runner

@joshmoore is this relevant???

https://github.blog/changelog/2024-01-30-github-actions-introducing-the-new-m1-macos-runner-available-to-open-source/

joshmoore commented:
I think so! This is what we've been waiting for.

joshmoore committed:
Revert generic change

joshmoore committed:
Add M1 macos runner (-14)

see:
 - https://github.blog/changelog/2024-01-30-github-actions-introducing-the-new-m1-macos-runner-available-to-open-source/
 - https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners/about-github-hosted-runners

joshmoore committed:
Try bumping setup-conda

joshmoore committed:
Try conda-incubator

joshmoore committed:
Add 'miniconda-version: latest'

joshmoore committed:
Bump bioformats2raw-libs as well

joshmoore committed:
Remove strict version dependency

joshmoore commented:
Hmmm..... well, it's green. Thoughts?

One from my side is that we should really just be building the noarch part once, but I'll leave that for another day.

manics commented on bioformats2raw/meta.yaml:
```diff
@@ -6,18 +6,16 @@ package:
-  - m2-patch    # [win]
   run:
-  - bioformats2raw-libs={{ PKG_VERSION }}-{{ PKG_BUILDNUM }}
+  - bioformats2raw-libs
```
The version pin is needed to ensure `conda upgrade bioformats2raw` automatically upgrades the `-libs` package in sync.

joshmoore replied:
Fair enough. Building them together, locally, though is no longer working. Any ideas on how to fix that? None of the following worked:

* `-c local`
* `--use-local`
* `conda build biofornats2raw{-libs,}` (i.e. at the same time)
* `... --output-folder X` (i.e. into the same folder)

manics replied:
Dunno, maybe you need this?
```
    defaults:
      run:
        shell: bash -el {0}
```
I've no idea why but it's in the README:
https://github.com/conda-incubator/setup-miniconda?tab=readme-ov-file#usage-examples

joshmoore replied:
It's failing even locally just with `conda build`.

joshmoore replied:
Would the use of `outputs:` be a more modern version of what's going on here?

e.g., https://github.com/AnacondaRecipes/boost-feedstock/blob/master/recipe/meta.yaml

joshmoore committed:
Try adding build requirement

joshmoore committed:
Use -c local and a single command

joshmoore committed:
Try croot

joshmoore committed:
Combine the packages

joshmoore committed:
Bump to 0.9.1

joshmoore committed:
Add c-blosc dependency

joshmoore committed:
Fix test

joshmoore commented:
I've failed all kinds of ways to get the two package version to build. I've remerged them.

joshmoore committed:
Fix build

joshmoore committed:
Bump to upload/download-artifact v4

joshmoore committed:
Set upload name

joshmoore committed:
Fix upload glob

joshmoore commented:
Build is now green. I'm going to merge and tag with `rc` so I can test with a `--pre` flag, then we should be ready for the final sign off on this.

jni commented:
Cool! Lemme know if you need some testers! 😊
